src/Components/Publishing/Display/DisplayPanel.tsx

```diff
@@ -48,6 +49,55 @@ export class DisplayPanel extends Component<Props, State> {
     }
   }
 
+  componentWillUpdate() {
+    if (this.video) {
+      this.video.removeEventListener("timeupdate", this.trackProgress)
+    }
+  }
+
+  componentDidUpdate() {
+    if (this.video) {
+      this.video.addEventListener("timeupdate", this.trackProgress)
+    }
+  }
+
+  // TODO: This could be shared with <CanvasVideo />
+  trackProgress = () => {
+    const secondsComplete = Math.floor(this.video.currentTime)
+    const percentComplete = Math.floor(this.video.currentTime / this.video.duration * 100)
+    const percentCompleteInterval = Math.floor(percentComplete / 25) * 25
+
+    // Track 25% duration intervals
+    if (percentCompleteInterval > 0) {
+      this.trackDuration(percentCompleteInterval)
+    }
+
+    // Track 3 & 10 seconds
+    if (secondsComplete === 3 || secondsComplete === 10) {
+      this.trackSeconds(secondsComplete)
+    }
+  }
+
+  trackDuration = memoize((percentComplete) => {
+    this.props.tracking.trackEvent({
+      action: "Video duration",
+      label: "Display ad video duration",
+      percent_complete: percentComplete,
+      campaign_name: this.props.campaign.name,
+      unit_layout: "panel"
+    })
+  })
+
+  trackSeconds = memoize((secondsComplete) => {
```

Comment:
@damassi I'm curious why this memoize is considered an optimization? If the arguments are always different, would the cached method ever be used twice?

Reply:
This is a trick I used in the canvas component to avoid accidentally double tracking the same 3 or 10 second intervals. Notice how [we round up](https://github.com/artsy/reaction/pull/370/files#diff-cbef750817163c85f91a3113eb1ca059R66) above—this could cause 3:05, 3:20, 3:35, etc. to all track 3 seconds.

Reply:
Ohh ok that makes sense with the rounding up. Wow that's cool! So on the first call, memoize applies the function and caches the result, then on subsequent calls it just returns the cached result? Verrrrry nifty!!